pkgs/development/cuda-modules/has-cuda-objects.nix

```diff
@@ -0,0 +1,105 @@
+{
+  lib,
+  config,
+  runCommand,
+  cudaPackages,
+  fd,
+}:
+{
+  package,
+  hasPTX ? false,
+  requiredCapabilities ? [],
+  preferLocalBuild ? true,
+  ...
+}@args:
+
+assert lib.assertMsg (lib.isDerivation package)
+  "provided package is not a derivation";
+
+# `find` will search the whole store if this list is empty
+assert lib.assertMsg (package.outputs != [])
+  "provided package has no outputs";
+
+assert lib.assertMsg (args.__structuredAttrs or true)
+  "This test requires __structuredAttrs to be true";
+
+let
+
+  args' = lib.removeAttrs args [ "package" ];
+
+  inherit (cudaPackages.flags) dropDot;
+
+in
+
+runCommand "test-any-cuobjdump-${package.name}"
+  (
+    args'
+    // {
+      inherit preferLocalBuild;
+      nativeBuildInputs = args'.nativeBuildInputs or [ ] ++ [
+        cudaPackages.cuda_cuobjdump
+      ];
+
+      __structuredAttrs = true;
+      package_outputs = lib.forEach package.outputs (output: package.${output});
+    }
+  )
+  ''
+    #set -x
+    any_cuobj_found=false
+    while IFS= read -r -d "" fname; do
+      echo "Checking $fname"
+
+      failed=false
+      cuobjdump_output=$(cuobjdump "$fname" 2>&1 || failed=true)
+
+      if ! $failed && ! grep <<<"$cuobjdump_output" -q "does not contain device code"; then
+        echo "Found cuda objects in $fname"
+        any_cuobj_found=true
+
+        # we could `continue`, but this is way better when debugging
+        all_requirements_found=true
+
+        # TODO: does this work?
+        ${lib.optionalString hasPTX ''
+          # we run cuobjdump again since --list-ptx interferes with requiredCapabilities
+          if cuobjdump --list-ptx "$fname" 1>/dev/null 2>/dev/null; then
+            echo "  Has PTX"
+          else
+            echo "  Does NOT have PTX"
+            all_requirements_found=false
+          fi
+        ''}
```

Comment:
how can i test for runtime compiled PTX objects?

Reply:
You mean what packages you can test this tester against?

Reply:
Yes. Currently the test in its current form only checks the exit code of cuobjdump, but for librealsense it prints:

```
PTX file    1: librealsense2.so.2.56.1.sm_52.ptx
PTX file    2: librealsense2.so.2.56.2.sm_52.ptx
PTX file    3: librealsense2.so.2.56.3.sm_52.ptx
```

which targets sm_52, and thus I assume it is not runtime compiled.

Reply:
Sorry for the delay. I meant to suggest that we prepare a tiny test derivation that invokes nvcc manually with well controlled gencodes. Then we generate a bunch of test and tester derivations from the same inputs, ensuring synchronization